cloud-controller-manager/do/loadbalancers.go

```diff
@@ -277,25 +282,92 @@ func (l *loadBalancers) UpdateLoadBalancer(ctx context.Context, clusterName stri
 func (l *loadBalancers) EnsureLoadBalancerDeleted(ctx context.Context, clusterName string, service *v1.Service) error {
-	lbName := l.GetLoadBalancerName(ctx, clusterName, service)
-	lb, found := l.resources.LoadBalancerByName(lbName)
-	if !found {
-		return nil
+	// Not calling retrieveAndAnnotateLoadBalancer to save a potential PATCH API
+	// call: the load-balancer is destined to be removed anyway.
+	lb, err := l.retrieveLoadBalancer(ctx, service)
+	if err != nil {
+		if err == errLBNotFound {
+			return nil
+		}
+		return err
 	}
 
 	resp, err := l.client.LoadBalancers.Delete(ctx, lb.ID)
 	if err != nil {
 		if resp != nil && resp.StatusCode == http.StatusNotFound {
-			l.resources.DeleteLoadBalancer(*lb)
 			return nil
 		}
 		return fmt.Errorf("failed to delete load-balancer: %s", err)
 	}
 
-	l.resources.DeleteLoadBalancer(*lb)
 	return nil
 }
 
+func (l *loadBalancers) retrieveAndAnnotateLoadBalancer(ctx context.Context, service *v1.Service) (*godo.LoadBalancer, error) {
+	lb, err := l.retrieveLoadBalancer(ctx, service)
+	if err != nil {
+		// Return bare error to easily compare for errLBNotFound. Converting to
+		// a full error type doesn't seem worth it.
+		return nil, err
+	}
+
+	if err := l.ensureLoadBalancerIDAnnot(service, lb.ID); err != nil {
+		return nil, fmt.Errorf("failed to add load-balancer ID annotation to service %s/%s: %s", service.Namespace, service.Name, err)
+	}
+
+	return lb, nil
+}
+
+func (l *loadBalancers) retrieveLoadBalancer(ctx context.Context, service *v1.Service) (*godo.LoadBalancer, error) {
+	if id := getLoadBalancerID(service); id != "" {
+		klog.V(2).Infof("Looking up load-balancer for service %s/%s by ID %s", service.Namespace, service.Name, id)
+		lb, resp, err := l.client.LoadBalancers.Get(ctx, id)
+		if err != nil {
+			if resp != nil && resp.StatusCode == http.StatusNotFound {
+				return nil, errLBNotFound
+			}
+			return nil, fmt.Errorf("failed to get load-balancer by ID %s: %s", id, err)
+		}
+
+		return lb, nil
+	}
+
+	// Retrieve by exhaustive iteration.
+	lbName := getDefaultLoadBalancerName(service)
+	klog.V(2).Infof("Looking up load-balancer for service %s/%s by name %s", service.Namespace, service.Name, lbName)
+	return l.lbByName(ctx, lbName)
+}
+
+func (l *loadBalancers) ensureLoadBalancerIDAnnot(service *v1.Service, lbID string) error {
+	if val := getLoadBalancerID(service); val != "" {
+		return nil
+	}
+
+	// Make a copy so we don't mutate the shared informer cache from the cloud
+	// provider framework.
+	updated := service.DeepCopy()
+	updated.ObjectMeta.Annotations[annoDOLoadBalancerID] = lbID
+
+	return patchService(l.resources.kclient, service, updated)
+}
+
+// lbByName gets a DigitalOcean Load Balancer by name. The returned error will
+// be lbNotFound if the load balancer does not exist.
+func (l *loadBalancers) lbByName(ctx context.Context, name string) (*godo.LoadBalancer, error) {
+	lbs, _, err := l.client.LoadBalancers.List(ctx, &godo.ListOptions{})
```

Comment:
I think this needs to paginate, like `allLoadBalancerList`

Reply:
Good catch!

Lucky for us, we already had a `allLoadBalancerList` function, so I just used that.